### ISSUE TITLE
[Bug]: 5 minutes to install the nuget package with a lot of package files

### ISSUE THREAD

**Issue by msedi** [NuGet/Home#12083]
### Impact

It's more difficult to complete my work

### Describe the bug

I'm creating an msbuild task which we deliver as a nuget package. Everything works fine so far, the task does what it should but a problem is the installation of the nuget package. It takes around 5 minutes to install the nuget package and there is not real log to identify the source of the lag.

During development I simply increase the version number to show a new package version. Afterwards I go to a demo project that contains the nuget package. I then simply update the nuget package in my demo project to the latest version, and this is what I get:

```
Restoring NuGet packages...
Running restore with 32 concurrent jobs.
Reading project file D:\GIT\IPIPE\Libraries\Managed\siemens.CUDA\siemens.CUDA.csproj.
The restore inputs for 'siemens.CUDA' have changed. Continuing restore.
Restoring packages for D:\GIT\IPIPE\Libraries\Managed\siemens.CUDA\siemens.CUDA.csproj...
Restoring packages for .NETCoreApp,Version=v6.0...
Resolving conflicts for net6.0...
Scanning packages for runtime.json files...
Restoring packages for net6.0/win-x64...
Resolving conflicts for net6.0/win-x64...
```

After around 5 minutes I get this output

```
Feeds used:
    https://api.nuget.org/v3/index.jsonC:\Program Files (x86)\Microsoft SDKs\NuGetPackages\
    D:\GIT\IPIPE\_Build\Release\Assemblies\ipipe.Cuda.Design
    Restored D:\GIT\IPIPE\Libraries\Managed\siemens.CUDA\siemens.CUDA.csproj (in 4,54 min).
NuGet package restore finished.
Time Elapsed: 00:04:32.6545900
```
Due to the transitivity of the project it takes more than 1 hour until my solution has restores (230 projects, but not all of them contains a relation to the package), but even if 10 projects contains a relation it takes around 50 minutes.

The problem is that I do not see who is wasting time. Is there anything I can do to get more logging?

### Repro Steps

For sure it has to do with my package and what I do, but there is no possibility to get some logs to identify the problem (?)

### Expected Behavior

In the end I expect a faster behavior but as said above I cannot get some logs to find out where it blocks.

**Comment by msedi** [comment]
Some more info on that. I'm distributing the package with ~1300 files. After removing them it works as before. The problem is that I need to distribute these files since they are part of the compiler. 

What I dont't understand what is taking so long because I have put the files in the tools folder and was assuming that the contents are not added somehow to the project the uses this nuget package.

**Comment by nkolev92** [comment]
Hey @msedi, 

Thanks for creating this issue. 

To help us understand your scenario better:

- Can you please provide a verbose output. It will have information about all the http calls made by NuGet. 
- Can you upload the assets file (note it can contain proprietary information). 
  The assets file can tell us how many packages were used, and whether we are dealing with a really large project graph.
- What's the size of the package in question? 

A package with a lot of files doesn't seem like a common NuGet scenario, so we might not be able to prioritize this in the foreseeable future.

**Comment by msedi** [comment]
@nkolev92. Thanks for getting back on this. Honestly I had this issue for at least a week now. But I made many changes. Just getting back to some of your question and then I will show what I did that might solved the issue.

> Can you please provide a verbose output. It will have information about all the http calls made by NuGet.
Unfortunately I have set the highest verbosity but it simply stops by saying ...resolving conflicts...

> What's the size of the package in question?
The size of the source nuget data is around 150MB, packaged its around 40MB

So now coming to a point that has maybe solved the problem. The package I produce is a design time only package, it only delivers two files (.props and .targets). Additionally I deliver a clang dll (~80MB) and CUDA NVRTC (~40MB). Both of them are packed in the `tools `directory setting the flag `<IsTool>true</IsTool>`.

Additionally I deliver many header files (*.h) required by LLVM/clang and CUDA itself. LLVM has no to much files (~20). But CUDA has more (~1300). Originally I had them in `content` and `contentFiles` because someone else in our team made it that way and I simply followed this. But there was never a problem before since I only used the LLVM headers, the CUDA files were added by me. It was never a problem before maybe due to the small amount of files. Since I added 1300 files with my changes, this seems to get a problem.

To solve this and because I don't need this header files as content at all, I simply distribute them also in the `tools` directory.

Honestyl even after that change it didn't solve it at first, but it seems that there was some caching somewhere so atfedr a few deletions and restarts of Visual Studio the problem was gone. I'm testing this for a few days now and never had problems again.

So in the end even if this is something you cannot follow to have this much contentFiles it is solved now by a different and maybe expected solution. The questions is it it really even makes sense to add 1300 files as content ;-)

**Comment by nkolev92** [comment]
Thanks for replying quickly. 

So just to confirm, after tweaking the package to move things into the `tools` folder and you don't have as many issues as before? 

https://docs.microsoft.com/en-us/nuget/concepts/package-installation-process is a great document that elaborates on NuGet package installations. If you do create a new package, normally you either need to clear that package from the global packages folder or create the package with a new version. It's extremely likely that you were hitting that. 
There's some obj/, bin/ level caching that you might have hit (not sure if it can cause the particular issues, just guessing), but probably not worth trying to figure it out now. 


If you don't have the really bad performance when the files are in the tools folder, then we might mark this as a duplicate of the following issue: https://github.com/NuGet/Home/issues/11543

**Comment by msedi** [comment]
Yes, I did all of the above, and moving it to `tools` made it work. I cleared the cache, removed all obj/bin (I already have a script for that in our projects, because we are having this issues more often). But as you maybe know when you working for a longer time on a problem you sometimes forget what you have done to make it work (or not) ;-)

So after all I think #11543 covers my original probem and we can close this issue.

Thanks for your help!
